### ISSUE TITLE
I referred to #28, but the test result is like this, can you tell me the reason? Thank you

### ISSUE THREAD

**Issue by saisai1002** [hzxie/Pix2Vox#51]
```
encoder = Encoder(cfg)
decoder = Decoder(cfg)
refiner = Refiner(cfg)
merger = Merger(cfg)

cfg.CONST.WEIGHTS = '/home/baijinggroup/lss/Pix2Vox/pretrained_weights/Pix2Vox-A-ShapeNet.pth'

checkpoint = torch.load(cfg.CONST.WEIGHTS, map_location=torch.device('cpu'))

fix_checkpoint = {}
fix_checkpoint['encoder_state_dict'] = OrderedDict((k.split('module.')[1:][0], v) for k, v in checkpoint['encoder_state_dict'].items())
fix_checkpoint['decoder_state_dict'] = OrderedDict((k.split('module.')[1:][0], v) for k, v in checkpoint['decoder_state_dict'].items())

epoch_idx = checkpoint['epoch_idx']
encoder.load_state_dict(fix_checkpoint['encoder_state_dict'])
decoder.load_state_dict(fix_checkpoint['decoder_state_dict'])

encoder.eval()
decoder.eval()
refiner.eval()
merger.eval()

img1_path = '/home/baijinggroup/lss/Pix2Vox/datasets/ShapeNetRendering/02691156/1a04e3eab45ca15dd86060f189eb133/rendering/00.png'
#img1_path = '/home/baijinggroup/lss/Pix2Vox/datasets/ShapeNetRendering/04090263/1aa5498ac780331f782611375da5ea9a/rendering/00.png'	
img1_np = np.asarray(Image.open(img1_path))
#img1_np = cv2.imread(img1_path)

sample = np.array([img1_np])

IMG_SIZE = cfg.CONST.IMG_H, cfg.CONST.IMG_W
CROP_SIZE = cfg.CONST.CROP_IMG_H, cfg.CONST.CROP_IMG_W

test_transforms = utils.data_transforms.Compose([
    utils.data_transforms.CenterCrop(IMG_SIZE, CROP_SIZE),
    utils.data_transforms.RandomBackground(cfg.TEST.RANDOM_BG_COLOR_RANGE), 
    utils.data_transforms.Normalize(mean=cfg.DATASET.MEAN, std=cfg.DATASET.STD),
    utils.data_transforms.ToTensor(),
])

rendering_images = test_transforms(rendering_images=sample)
rendering_images = rendering_images.unsqueeze(0)

with torch.no_grad():
    image_features = encoder(rendering_images)
    raw_features, generated_volume = decoder(image_features)

    if cfg.NETWORK.USE_MERGER and epoch_idx >= cfg.TRAIN.EPOCH_START_USE_MERGER:
        generated_volume = merger(raw_features, generated_volume)
    else:
        generated_volume = torch.mean(generated_volume, dim=1)

    if cfg.NETWORK.USE_REFINER and epoch_idx >= cfg.TRAIN.EPOCH_START_USE_REFINER:
        generated_volume = refiner(generated_volume)

generated_volume = generated_volume.squeeze(0)
img_dir = '/home/baijinggroup/lss/Pix2Vox/core/sample_images'
gv = generated_volume.cpu().numpy()
gv_new = np.swapaxes(gv, 2, 1)
rendering_views = utils.binvox_visualization.get_volume_views(gv_new, os.path.join(img_dir), epoch_idx)
```

This is the input:
![image](https://user-images.githubusercontent.com/23035261/95011110-f9a68900-0660-11eb-9c6f-984d686f53b4.png)

This is the model's output:
![image](https://user-images.githubusercontent.com/23035261/95011123-0b882c00-0661-11eb-97ea-10cb8c21152c.png)
The result after I added this code np.swapaxes(gv,2,1)：
![image](https://user-images.githubusercontent.com/23035261/95011164-57d36c00-0661-11eb-8e88-199c8b3e0704.png)

**Comment by brian220** [comment]
I got a better result by using `cv2.imread(img1_path, cv2.IMREAD_UNCHANGED).astype(np.float32) / 255.` to read the image. (Same as the author did in dataloader)

![voxels-000153](https://user-images.githubusercontent.com/27956674/95169446-782c3380-07e5-11eb-8708-ac113fd7b717.png)

Code:
`
import json
import numpy as np
import os
import torch
import torch.backends.cudnn
import torch.utils.data
import cv2
from collections import OrderedDict

from PIL import Image

import utils.binvox_visualization
import utils.data_loaders
import utils.data_transforms
import utils.network_utils

from datetime import datetime as dt

from models.encoder import Encoder
from models.decoder import Decoder
from models.refiner import Refiner
from models.merger import Merger

def test_single_img_net(cfg):

    # Enable the inbuilt cudnn auto-tuner to find the best algorithm to use
    torch.backends.cudnn.benchmark = True
    
    encoder = Encoder(cfg)
    decoder = Decoder(cfg)
    refiner = Refiner(cfg)
    merger = Merger(cfg)
    
    cfg.CONST.WEIGHTS = '/home/caig/Desktop/pix2vox/Pix2Vox/pretrain/Pix2Vox-A-ShapeNet.pth'
    checkpoint = torch.load(cfg.CONST.WEIGHTS, map_location=torch.device('cpu'))
    
    fix_checkpoint = {}
    fix_checkpoint['encoder_state_dict'] = OrderedDict((k.split('module.')[1:][0], v) for k, v in checkpoint['encoder_state_dict'].items())
    fix_checkpoint['decoder_state_dict'] = OrderedDict((k.split('module.')[1:][0], v) for k, v in checkpoint['decoder_state_dict'].items())
    fix_checkpoint['refiner_state_dict'] = OrderedDict((k.split('module.')[1:][0], v) for k, v in checkpoint['refiner_state_dict'].items())
    fix_checkpoint['merger_state_dict'] = OrderedDict((k.split('module.')[1:][0], v) for k, v in checkpoint['merger_state_dict'].items())
    
    epoch_idx = checkpoint['epoch_idx']
    encoder.load_state_dict(fix_checkpoint['encoder_state_dict'])
    decoder.load_state_dict(fix_checkpoint['decoder_state_dict'])

    if cfg.NETWORK.USE_REFINER:
        print('Use refiner')
        refiner.load_state_dict(fix_checkpoint['refiner_state_dict'])
    if cfg.NETWORK.USE_MERGER:
        print('Use merger')
        merger.load_state_dict(fix_checkpoint['merger_state_dict'])

    
    encoder.eval()
    decoder.eval()
    refiner.eval()
    merger.eval()
    
    img1_path = '/home/caig/Desktop/data/shapenet/ShapeNetRendering/02691156/1a04e3eab45ca15dd86060f189eb133/rendering/00.png'
    # img1_path = '/home/caig/Desktop/data/shapenet/ShapeNetRendering/04379243/1a1fb603583ce36fc3bd24f986301745/rendering/00.png'
    # img1_np = np.asarray(Image.open(img1_path))
    img1_np = cv2.imread(img1_path, cv2.IMREAD_UNCHANGED).astype(np.float32) / 255.
    
    sample = np.array([img1_np])
    
    IMG_SIZE = cfg.CONST.IMG_H, cfg.CONST.IMG_W
    CROP_SIZE = cfg.CONST.CROP_IMG_H, cfg.CONST.CROP_IMG_W
    
    test_transforms = utils.data_transforms.Compose([
        utils.data_transforms.CenterCrop(IMG_SIZE, CROP_SIZE),
        utils.data_transforms.RandomBackground(cfg.TEST.RANDOM_BG_COLOR_RANGE),
        utils.data_transforms.Normalize(mean=cfg.DATASET.MEAN, std=cfg.DATASET.STD),
        utils.data_transforms.ToTensor(),
    ])
    
    rendering_images = test_transforms(rendering_images=sample)
    rendering_images = rendering_images.unsqueeze(0)
    
    with torch.no_grad():
        image_features = encoder(rendering_images)
        raw_features, generated_volume = decoder(image_features)
    
        if cfg.NETWORK.USE_MERGER and epoch_idx >= cfg.TRAIN.EPOCH_START_USE_MERGER:
            generated_volume = merger(raw_features, generated_volume)
        else:
            generated_volume = torch.mean(generated_volume, dim=1)
    
        if cfg.NETWORK.USE_REFINER and epoch_idx >= cfg.TRAIN.EPOCH_START_USE_REFINER:
            generated_volume = refiner(generated_volume)

    generated_volume = generated_volume.squeeze(0)

    img_dir = '/home/caig/Desktop/pix2vox/Pix2Vox/outputs'
    gv = generated_volume.cpu().numpy()
    gv_new = np.swapaxes(gv, 2, 1)
    rendering_views = utils.binvox_visualization.get_volume_views(gv_new, os.path.join(img_dir),
                                                                                        epoch_idx)
`

**Comment by lui-shex** [comment]
where were you to add this code in this project?

Thanks for response!

**Comment by brian220** [comment]
> where were you to add this code in this project?
> 
> Thanks for response!

I add this code in the core folder and give it a name (ex. test_single_img.py)
then modify the runner.py to run it

![image](https://user-images.githubusercontent.com/27956674/97105838-ee42fc80-16f8-11eb-88a9-ef9348901c0c.png)

Change line 77 here to the class name in the new file (ex. in test_single_img.py)

(Please check test.py and runner.py for more details)

Thanks!

**Comment by Zc132585** [comment]
> I got a better result by using `cv2.imread(img1_path, cv2.IMREAD_UNCHANGED).astype(np.float32) / 255.` to read the image. (Same as the author did in dataloader)我用来 `cv2.imread(img1_path, cv2.IMREAD_UNCHANGED).astype(np.float32) / 255.` 读取图片得到了更好的结果。（和作者在 dataloader 里做的一样）
> 
> <img alt="voxels-000153" width="388.9961853027344" height="388.9961853027344" src="https://user-images.githubusercontent.com/27956674/95169446-782c3380-07e5-11eb-8708-ac113fd7b717.png">
> 
> Code:  代码：
> 
> ```
>  encoder = Encoder(cfg)
>  decoder = Decoder(cfg)
>  refiner = Refiner(cfg)
>  merger = Merger(cfg)
> 
>  cfg.CONST.WEIGHTS = '/home/caig/Desktop/pix2vox/Pix2Vox/pretrain/Pix2Vox-A-ShapeNet.pth'
>  checkpoint = torch.load(cfg.CONST.WEIGHTS, map_location=torch.device('cpu'))
>  
>  fix_checkpoint = {}
>  fix_checkpoint['encoder_state_dict'] = OrderedDict((k.split('module.')[1:][0], v) for k, v in checkpoint['encoder_state_dict'].items())
>  fix_checkpoint['decoder_state_dict'] = OrderedDict((k.split('module.')[1:][0], v) for k, v in checkpoint['decoder_state_dict'].items())
>  fix_checkpoint['refiner_state_dict'] = OrderedDict((k.split('module.')[1:][0], v) for k, v in checkpoint['refiner_state_dict'].items())
>  fix_checkpoint['merger_state_dict'] = OrderedDict((k.split('module.')[1:][0], v) for k, v in checkpoint['merger_state_dict'].items())
>  
>  epoch_idx = checkpoint['epoch_idx']
>  encoder.load_state_dict(fix_checkpoint['encoder_state_dict'])
>  decoder.load_state_dict(fix_checkpoint['decoder_state_dict'])
> 
>  if cfg.NETWORK.USE_REFINER:
>      print('Use refiner')
>      refiner.load_state_dict(fix_checkpoint['refiner_state_dict'])
>  if cfg.NETWORK.USE_MERGER:
>      print('Use merger')
>      merger.load_state_dict(fix_checkpoint['merger_state_dict'])
> 
>  
>  encoder.eval()
>  decoder.eval()
>  refiner.eval()
>  merger.eval()
>  
>  img1_path = '/home/caig/Desktop/data/shapenet/ShapeNetRendering/02691156/1a04e3eab45ca15dd86060f189eb133/rendering/00.png'
>  img1_np = cv2.imread(img1_path, cv2.IMREAD_UNCHANGED).astype(np.float32) / 255.
>  
>  sample = np.array([img1_np])
>  
>  IMG_SIZE = cfg.CONST.IMG_H, cfg.CONST.IMG_W
>  CROP_SIZE = cfg.CONST.CROP_IMG_H, cfg.CONST.CROP_IMG_W
>  
>  test_transforms = utils.data_transforms.Compose([
>      utils.data_transforms.CenterCrop(IMG_SIZE, CROP_SIZE),
>      utils.data_transforms.RandomBackground(cfg.TEST.RANDOM_BG_COLOR_RANGE),
>      utils.data_transforms.Normalize(mean=cfg.DATASET.MEAN, std=cfg.DATASET.STD),
>      utils.data_transforms.ToTensor(),
>  ])
>  
>  rendering_images = test_transforms(rendering_images=sample)
>  rendering_images = rendering_images.unsqueeze(0)
>  
>  with torch.no_grad():
>      image_features = encoder(rendering_images)
>      raw_features, generated_volume = decoder(image_features)
>  
>      if cfg.NETWORK.USE_MERGER and epoch_idx >= cfg.TRAIN.EPOCH_START_USE_MERGER:
>          generated_volume = merger(raw_features, generated_volume)
>      else:
>          generated_volume = torch.mean(generated_volume, dim=1)
>  
>      if cfg.NETWORK.USE_REFINER and epoch_idx >= cfg.TRAIN.EPOCH_START_USE_REFINER:
>          generated_volume = refiner(generated_volume)
> 
>  generated_volume = generated_volume.squeeze(0)
> 
>  img_dir = '/home/caig/Desktop/pix2vox/Pix2Vox/outputs'
>  gv = generated_volume.cpu().numpy()
>  gv_new = np.swapaxes(gv, 2, 1)
>  rendering_views = utils.binvox_visualization.get_volume_views(gv_new, os.path.join(img_dir),
>                                                                                      epoch_idx)
> ```

你好，可以详细说说你是怎么复现单张图重建的吗，我写了一份推理代码但复现的结果很差，你方便提供一份完整的推理代码吗，同时你说的需要修改这里的runner.py的代码吗？多张图的推理代码有吗，十分感谢
